### PR TITLE
Optimize link cell in case where cutoff is close to cell width.

### PR DESCRIPTION
**Pull request by vyasr**
This applies the optimization we discussed a while ago for LinkCell. @jinsoo960 can you quickly test whether this recovers the expected performance? If so, it should be fine to merge.

### CODE REVIEW

File: cpp/locality/LinkCell.h [hunk 1]
```diff
@@ -538,7 +538,19 @@ class LinkCellQueryBallIterator : virtual public LinkCellIterator
                               float r, bool exclude_ii)
         : NeighborQueryIterator(neighbor_query, points, N, exclude_ii),
           LinkCellIterator(neighbor_query, points, N, exclude_ii), m_r(r)
-    {}
+    {
+        // Upon querying, if the search radius is equal to the cell width, we
+        // can guarantee that we don't need to search the cell shell past the
+        // query radius. For simplicity, we store this value as an integer.
+        if (abs(m_r - neighbor_query->getCellWidth()) < 1e-6)
```
Comment (bdice) [inline cpp/locality/LinkCell.h]: @vyasr Why not `m_r == neighbor_query->getCellWidth()`? Or at least use a relative tolerance -- I'm hesitant to introduce a scale-dependent parameter like this.

Comment (vyasr) [reply cpp/locality/LinkCell.h]: That's fair, checking for equality is probably fine. If for whatever reason the simple equality check fails due to some imprecision we're better off falling back to the slower but guaranteed correct version.

**Comment by jinsooihm** [comment]
It is twice faster than calling `queryBall(...).toNList()` on the original implementation, but still twice slower compared to `compute(...)`. Hopefully, this difference should diminish once we change the order of `ref_points` and `points` as we just discussed @vyasr.

**Comment by vyasr** [comment]
To provide a bit more information: when we originally benchmarked, there were two main slowdowns relative to the original `compute` method. The first was that many extra cells were being searched in the new method. This pull request addresses that problem, because in the specific use-case that the `compute` method supports, namely when the rcut is equal to the cell width, you can stop searching one cell earlier than you have to in the more general case. However, this pull request does not address the other reason for the slowdown, which is that the sorting to generate the NeighborList is slower in the `query...toNeighborList` approach  in the `compute` method. The reason for this is that the NeighborList is ordered by `ref_point, point`, and `compute` is designed around this order, whereas in freud 2.0 we are swapping that ordering and the `query*` methods are designed around that ordering. Once we perform the swap (in PR #342 into the `next` branch), these performance characteristics should be reversed and we should recover the original performance.

**Review by jinsooihm — APPROVED** [review]
Looks good.

**Review by bdice — APPROVED** [review]
👍

**Pull request merged**